bubblewrap.c

```diff
@@ -3103,6 +3112,35 @@ main (int    argc,
                          -1, FALSE, FALSE);
     }
 
+  if (opt_disable_userns)
+    {
+      cleanup_fd int sysctl_fd = -1;
+
+      sysctl_fd = openat (proc_fd, "sys/user/max_user_namespaces", O_WRONLY);
+
+      if (sysctl_fd < 0)
+        die_with_error ("cannot open /proc/sys/user/max_user_namespaces");
+
+      if (write_to_fd (sysctl_fd, "1", 1) < 0)
+        die_with_error ("sysctl user.max_user_namespaces = 1");
+
+      if (unshare (CLONE_NEWUSER))
+        die_with_error ("unshare user ns");
+
+      /* Check that we can't make a new userns again */
```

Comment:
might as well `sysctl user.max_user_namespaces = 2` and then try to create the user ns

Reply:
> might as well sysctl user.max_user_namespaces = 2 and then try to create the user ns

This seems like it would be misleading, because setting the sysctl to a higher value succeeds and the unshare still fails, so I think that would give a reader the wrong idea about why this works. To verify this, I patched `bubblewrap.c` like so:

```diff
diff --git a/bubblewrap.c b/bubblewrap.c
index cd57e86..8e21fc2 100644
--- a/bubblewrap.c
+++ b/bubblewrap.c
@@ -3200,6 +3200,8 @@ main (int    argc,
 
           if (write_to_fd (sysctl_fd, "1", 1) < 0)
             die_with_error ("sysctl user.max_user_namespaces = 1");
+
+          fprintf (stderr, "max userns set to 1\n");
         }
 
       if (unshare (CLONE_NEWUSER))
@@ -3216,10 +3218,31 @@ main (int    argc,
 
   if (opt_disable_userns || opt_assert_userns_disabled)
     {
+      cleanup_fd int sysctl_fd = -1;
+
+      sysctl_fd = openat (proc_fd, "sys/user/max_user_namespaces", O_WRONLY);
+
+      if (sysctl_fd < 0)
+        die_with_error ("cannot open /proc/sys/user/max_user_namespaces");
+
+      if (write_to_fd (sysctl_fd, "42", 2) < 0)
+        die_with_error ("sysctl user.max_user_namespaces = 42");
+
+      fprintf (stderr, "max userns set to 42\n");
+
       /* Verify that we can't make a new userns again */
       res = unshare (CLONE_NEWUSER);
       if (res == 0)
         die ("unable to disable creation of new user namespaces");
+      fprintf (stderr, "unshare(CLONE_NEWUSER) failed as expected, with: %s\n", strerror (errno));
+
+#if 0
+      (void) lseek (sysctl_fd, 0, SEEK_SET);
+      if (write_to_fd (sysctl_fd, "1", 1) < 0)
+        die_with_error ("sysctl user.max_user_namespaces = 1");
+
+      fprintf (stderr, "max userns back to 1\n");
+#endif
     }
 
   /* All privileged ops are done now, so drop caps we don't need */
```

With or without the block that's in `#if 0` here, everything here succeeds, except for the `unshare()` which fails with `ENOSPC` as we would hope. What is happening here (as I understand it) is that the bubblewrap process has capabilities in the userns that it is creating or joining (because we haven't called `drop_privs()` yet), like so:

* initns, owned by uid 0, limit set to 10000 or something
    * intermediate ns, owned by us, limit set to 1
        * final ns, owned by us, limit set to 42

and so we are allowed to set the limit to anything we want to, even if it's greater than the parent namespace's limit. However (as documented in `namespaces(7)`), when we try to create a new userns, the resulting number of user namespaces is checked against "the corresponding namespace limit for the creator UID in the ancestor namespace", and if that limit would be exceeded then creation fails.

So the `user.max_user_namespaces` in the final namespace is actually largely irrelevant, and it's the `user.max_user_namespaces` in the intermediate namespace (in which members of the final namespace are incapable) which matters.

I don't really want to set the limit to a higher value for this check and then leave it there, because it seems unexpected that a process could read out `user.max_user_namespaces`, think "oh that's fine, I'm probably allowed to create up to 41 more namespaces", and then find that it gets `ENOSPC`.

I also don't really want to set the limit higher and then reset it back down to 1 (as would happen if the `#if 0` block is uncommented), because if I do that, it looks like a security-significant action even though actually it's not. Also, it'll look misleading in `strace` or similar.

And, I think it's possible to enter the `opt_assert_userns_disabled` block even if we're setuid root, which means we need to be careful not to do anything that could be a net user privilege increase (like increasing the `user.max_user_namespaces` of the init namespace).